projects/jupyter-server-ydoc/jupyter_server_ydoc/handlers.py

```diff
@@ -459,3 +463,141 @@ async def put(self, path):
         )
         self.set_status(201)
         return self.finish(data)
+
+
+class TimelineHandler(APIHandler):
+    def initialize(
+        self, ystore_class: type[BaseYStore], ywebsocket_server: JupyterWebsocketServer
+    ) -> None:
+        self.ystore_class = ystore_class
+        self.ywebsocket_server = ywebsocket_server
+
+    async def get(self, path: str) -> None:
+        idx = uuid4().hex
+        file_id_manager = self.settings["file_id_manager"]
+        file_id = file_id_manager.get_id(path)
+
+        format = str(self.request.query_arguments.get("format")[0].decode("utf-8"))
+        content_type = str(self.request.query_arguments.get("type")[0].decode("utf-8"))
+        encoded_path = encode_file_path(format, content_type, file_id)
+        try:
+            room_id = room_id_from_encoded_path(encoded_path)
+            room: YRoom = await self.ywebsocket_server.get_room(room_id)
+            fork_ydoc = Doc()
+
+            ydoc_factory = YDOCS.get(content_type)
+            if ydoc_factory is None:
+                self.set_status(404)
+                self.finish(
+                    {
+                        "code": 404,
+                        "error": f"No document factory found for content type: {content_type}",
+                    }
+                )
+                return
+
+            FORK_DOCUMENTS[idx] = ydoc_factory(fork_ydoc)
+            undo_manager: UndoManager = FORK_DOCUMENTS[idx].undo_manager
+
+            updates_and_timestamps = [(item[0], item[-1]) async for item in room.ystore.read()]
+
+            result_timestamps = []
+
+            for update, timestamp in updates_and_timestamps:
+                undo_stack_len = len(undo_manager.undo_stack)
+                fork_ydoc.apply_update(update)
+                if len(undo_manager.undo_stack) > undo_stack_len:
+                    result_timestamps.append(timestamp)
+
+            fork_room = YRoom()
+            fork_room.ydoc = fork_ydoc
```

Comment:
With `pycrdt-websocket` v0.14.2, this can be replaced with:
```suggestion
            fork_room = YRoom(fork_ydoc)
```